Quiz\Crossword: Multiple Tries clears the letter at the intersection …

BruceGoodGuy opened the pull request:
Hi @timhunt ,
In this pull request I fixed 2 issues:

Issue 1: Problem with multiple attempts with clearing incorrect response settings.
Currently, when the user attempts the crossword with multiple tries and clears the incorrect response setting.
If one answer is wrong and the user clicks retry, all wrong answers will be clear and it will remove the letter of the correct answer (if they are at the intersection). (please refer to ticket Defect [624230](https://dev.azure.com/openuniversity/IT/_workitems/edit/624230) for more details)

Issue 2: Problem with whitespace.
When a user enters a space at the beginning of an answer, every time they submit, all answers will be truncated with spaces
at the beginning and at the end. For example, if the user enters 2 spaces in the answer: "   LLO ", the answer will be removed 2 whitespaces when submitting: "LLO".

Could you please help me review it? Thanks.

timhunt commented:
I have only looked at this quickly, but some initial comments?

1. Why the issue with spaces? I thought that we used _ characters when a cell was blank. Where did the spaces come from?
2. Please read https://moodledev.io/general/development/abc/comments. I think `@param {Boolean} skipEmptyData Skip empty data flag.` is an example of exactly what that pages says not to do, and I don't understand what that argument is diong.

timhunt commented:
Well, looking at the Behat test, perhaps I can guess the space thing.

So, up until now, we would expect the value in the input to be "__RIS", but rather than making users type _, we want them to be able to just press space. That seems like good usability.

But, that is not a good reason to change anything else. Why not, if they press space, insert a _ into the input with some client-side JS? That way, the input format is always consistent?

BruceGoodGuy commented:
Hi @timhunt,

As per your comment, I edited the comment for skipEmptyData.

Then I also changed the spacing input behavior. From now on, when the user types a space, we will insert an underscore character into the input instead of a space. I also changed the code in renderer.php, question.php to make sure the crossword works well with the new behavior.

Could you please help me to review it? thanks.
New commit url: [URL](https://github.com/moodleou/moodle-qtype_crossword/pull/8/commits/23b68f94ef8d2a27041d4de0a692a2231cc97fa1)

timhunt reviewed (changes requested):
Thank you for continuing to work on this. I think we are getting closer to a good solution, but I don't think we are there yet.

The problem I see with what you have done is, that to implement a single concept (replace _ with space before processing) you have had to make the same change in a lot of different places. That is a serious violation of the DRY principle (in the sense of https://pragprog.com/titles/tpp20/the-pragmatic-programmer-20th-anniversary-edition/ - recommended book, by the way.)

I wonder if, a better approach, would be ... Well, it seems that all we ever do with the response is to compare $answer->answer === $response[$this->field($key)]. Therefore, what about making a new method so this can be $answer->is_correct($response[$this->field($key)]), and the logic for handling _ and space can be in that method, and also, hopefully, it makes all of the code easier to read? What do you think?

timhunt commented on classes/util.php:
```diff
@@ -48,5 +48,17 @@ public static function safe_normalize(string $string): string {
+      * @param string $replacetext Specifies the value to replace the value.
+      * @return array The new array list.
+      */
+    public static function replace_underscore_from_array(array $data, string $replacetext = ' '): array {
```
Given the above comment, actually we may be going to delete this method, in which case ignore this.

timhunt commented on amd/src/crossword_question.js:
```diff
@@ -283,6 +283,10 @@ export class CrosswordQuestion {
+                    // Replace spaces with an underscore character before binding to the answer input.
+                    if (key === " ") {
+                        key = '_';
+                    }
```
Mixture of quotes is a bit sloppy. Can you consistently use ' ' unless you need " ".

timhunt commented on classes/answer.php:
```diff
@@ -57,4 +57,15 @@ public function __construct(string $answer, string $clue, string $orientation,
+     * @return bool The result after check, True if correct.
+     */
+    public function is_correct(string $answer): bool {
+        return $this->answer === str_replace('_', " ", $answer);
```
Mixed quotes again.

timhunt reviewed (changes requested):
I think the new code is better. I hope you agree.

Are you able to tidy up git history? (I think merge the second and third commit. you can do that with git rebase -i.)

And, since you are diong that, I pointed out two other small improvements that could be made.
